// @vidstack/react/dist/prod/vidstack.js
"use client"

import { D as DOMEvent, j as isString, l as listenEvent, L as createReactComponent, u as useStateContext, a as useSignal, C as composeRefs, b as useSignalRecord, g as effect, J as useReactScope, s as signal } from './chunks/vidstack-BNPvxaQh.js';
export { Q as appendTriggerEvent, P as findTriggerEvent, N as hasTriggerEvent, F as isKeyboardClick, E as isKeyboardEvent, R as isPointerEvent, O as walkTriggerEventChain } from './chunks/vidstack-BNPvxaQh.js';
import { R as GroupedLog, i as TextTrackSymbol, P as Primitive, a as MediaPlayerInstance, z as isRemotionProvider, U as MediaProviderInstance, V as mediaState, b as TextTrack, W as ToggleButtonInstance, X as PosterInstance, u as useMediaState, Y as ThumbnailsLoader, Z as updateSliderPreviewPlacement } from './chunks/vidstack-BRXztYRL.js';
export { a8 as ARIAKeyShortcuts, aV as AUDIO_EXTENSIONS, aW as AUDIO_TYPES, b5 as AirPlayButtonInstance, ak as AudioProviderLoader, aQ as AudioRadioGroup, af as AudioTrackList, a$ as CaptionButtonInstance, K as CaptionsInstance, aR as CaptionsRadioGroup, aP as ChaptersRadioGroup, aA as ControlsGroup, D as ControlsGroupInstance, C as ControlsInstance, aS as DEFAULT_PLAYBACK_RATES, b0 as FullscreenButtonInstance, a2 as FullscreenController, bl as GestureInstance, H as GoogleCastButtonInstance, al as HLSProviderLoader, aZ as HLS_VIDEO_EXTENSIONS, a_ as HLS_VIDEO_TYPES, a1 as List, b1 as LiveButtonInstance, a6 as LocalMediaStorage, a7 as MEDIA_KEY_SHORTCUTS, a5 as MediaControls, M as MediaRemoteControl, aJ as MenuButton, bf as MenuButtonInstance, be as MenuInstance, aK as MenuItem, bh as MenuItemInstance, aN as MenuItems, bg as MenuItemsInstance, aL as MenuPortal, bi as MenuPortalInstance, b2 as MuteButtonInstance, b3 as PIPButtonInstance, b4 as PlayButtonInstance, aU as QualityRadioGroup, aO as Radio, bj as RadioGroupInstance, bk as RadioInstance, a4 as ScreenOrientationController, b6 as SeekButtonInstance, aI as SliderChapters, bd as SliderChaptersInstance, aD as SliderController, S as SliderInstance, aH as SliderPreview, bc as SliderPreviewInstance, aE as SliderThumbnail, b9 as SliderThumbnailInstance, aG as SliderValue, ba as SliderValueInstance, aF as SliderVideo, bb as SliderVideoInstance, aT as SpeedRadioGroup, ac as TextRenderers, ae as TextTrackList, bm as ThumbnailInstance, bn as TimeInstance, T as TimeRange, b7 as TimeSliderInstance, aC as TooltipContent, G as TooltipContentInstance, E as TooltipInstance, aB as TooltipTrigger, F as TooltipTriggerInstance, aX as VIDEO_EXTENSIONS, aY as VIDEO_TYPES, am as VideoProviderLoader, aj as VideoQualityList, an as VimeoProviderLoader, b8 as VolumeSliderInstance, ao as YouTubeProviderLoader, _ as canChangeVolume, a3 as canFullscreen, f as canGoogleCastSrc, $ as canOrientScreen, y as canPlayHLSNatively, a0 as canRotateScreen, w as canUsePictureInPicture, x as canUseVideoPresentation, ag as findActiveCue, O as formatSpokenTime, N as formatTime, ab as getTimeRangesEnd, aa as getTimeRangesStart, ap as isAudioProvider, t as isAudioSrc, ah as isCueActive, au as isGoogleCastProvider, ar as isHLSProvider, m as isHLSSrc, av as isHTMLAudioElement, ay as isHTMLIFrameElement, ax as isHTMLMediaElement, aw as isHTMLVideoElement, o as isMediaStream, A as isTrackCaptionKind, aq as isVideoProvider, v as isVideoSrc, at as isVimeoProvider, as as isYouTubeProvider, B as mediaContext, aM as menuPortalContext, ad as parseJSONCaptionsFile, az as sliderContext, s as sliderState, a9 as softResetMediaState, bo as useMediaStore, r as usePlyrLayoutClasses, J as watchActiveTextTrack, ai as watchCueTextChange } from './chunks/vidstack-BRXztYRL.js';
import * as React from 'react';
import { u as useMediaContext } from './chunks/vidstack-D2C8-WOh.js';
export { A as AirPlayButton, C as CaptionButton, F as FullscreenButton, G as Gesture, L as LiveButton, E as Menu, M as MuteButton, b as PIPButton, P as PlayButton, D as RadioGroup, S as SeekButton, x as Slider, H as Thumbnail, e as Time, z as TimeSlider, y as VolumeSlider, h as useAudioOptions, l as useCaptionOptions, a as useMediaPlayer, k as usePlaybackRateOptions, m as useVideoQualityOptions } from './chunks/vidstack-D2C8-WOh.js';
import { I as Icon } from './chunks/vidstack-Xvq9wjeH.js';
import { l as createSignal, m as useScoped } from './chunks/vidstack-C023lPlt.js';
export { d as Captions, b as ChapterTitle, n as Controls, G as GoogleCastButton, s as Spinner, a as Title, t as Tooltip, p as useActiveTextCues, q as useActiveTextTrack, u as useChapterOptions, g as useChapterTitle, o as useTextCues } from './chunks/vidstack-C023lPlt.js';
export { u as useMediaRemote } from './chunks/vidstack-DYnY8p8l.js';
export { u as useSliderState, a as useSliderStore } from './chunks/vidstack-CPa7biz6.js';
import 'react-dom';

class Logger {
  constructor() {
    this.B = null;
  }
  error(...data) {
    return this.dispatch("error", ...data);
  }
  warn(...data) {
    return this.dispatch("warn", ...data);
  }
  info(...data) {
    return this.dispatch("info", ...data);
  }
  debug(...data) {
    return this.dispatch("debug", ...data);
  }
  errorGroup(title) {
    return new GroupedLog(this, "error", title);
  }
  warnGroup(title) {
    return new GroupedLog(this, "warn", title);
  }
  infoGroup(title) {
    return new GroupedLog(this, "info", title);
  }
  debugGroup(title) {
    return new GroupedLog(this, "debug", title);
  }
  setTarget(newTarget) {
    this.B = newTarget;
  }
  dispatch(level, ...data) {
    return this.B?.dispatchEvent(
      new DOMEvent("vds-log", {
        bubbles: true,
        composed: true,
        detail: { level, data }
      })
    ) || false;
  }
}

class LibASSTextRenderer {
  constructor(loader, config) {
    this.loader = loader;
    this.config = config;
    this.priority = 1;
    this.g = null;
    this.A = null;
    this.Ie = /(ssa|ass)$/;
  }
  canRender(track, video) {
    return !!video && !!track.src && (isString(track.type) && this.Ie.test(track.type) || this.Ie.test(track.src));
  }
  attach(video) {
    if (!video)
      return;
    this.loader().then(async (mod) => {
      this.g = new mod.default({
        ...this.config,
        video,
        subUrl: this.A?.src || ""
      });
      listenEvent(this.g, "ready", () => {
        const canvas = this.g?.Qj;
        if (canvas)
          canvas.style.pointerEvents = "none";
      });
      listenEvent(this.g, "error", (event) => {
        if (this.A) {
          this.A[TextTrackSymbol.N] = 3;
          this.A.dispatchEvent(
            new DOMEvent("error", {
              trigger: event,
              detail: event.error
            })
          );
        }
      });
    });
  }
  changeTrack(track) {
    if (!track || track.readyState === 3) {
      this.Je();
    } else if (this.A !== track) {
      this.g?.setTrackByUrl(track.src);
      this.A = track;
    }
  }
  detach() {
    this.Je();
  }
  Je() {
    this.g?.freeTrack();
    this.A = null;
  }
}

const playerCallbacks = [
  "onAbort",
  "onAudioTrackChange",
  "onAudioTracksChange",
  "onAutoPlay",
  "onAutoPlayChange",
  "onAutoPlayFail",
  "onCanLoad",
  "onCanPlay",
  "onCanPlayThrough",
  "onControlsChange",
  "onDestroy",
  "onDurationChange",
  "onEmptied",
  "onEnd",
  "onEnded",
  "onError",
  "onFindMediaPlayer",
  "onFullscreenChange",
  "onFullscreenError",
  "onLiveChange",
  "onLiveEdgeChange",
  "onLoadedData",
  "onLoadedMetadata",
  "onLoadStart",
  "onLoopChange",
  "onOrientationChange",
  "onPause",
  "onPictureInPictureChange",
  "onPictureInPictureError",
  "onPlay",
  "onPlayFail",
  "onPlaying",
  "onPlaysInlineChange",
  "onPosterChange",
  "onProgress",
  "onProviderChange",
  "onProviderLoaderChange",
  "onProviderSetup",
  "onQualitiesChange",
  "onQualityChange",
  "onRateChange",
  "onReplay",
  "onSeeked",
  "onSeeking",
  "onSourceChange",
  "onSourceChange",
  "onStalled",
  "onStarted",
  "onStreamTypeChange",
  "onSuspend",
  "onTextTrackChange",
  "onTextTracksChange",
  "onTimeUpdate",
  "onTitleChange",
  "onVdsLog",
  "onVideoPresentationChange",
  "onVolumeChange",
  "onWaiting"
];

const MediaPlayerBridge = createReactComponent(MediaPlayerInstance, {
  events: playerCallbacks,
  eventsRegex: /^onHls/,
  domEventsRegex: /^onMedia/
});
const MediaPlayer = React.forwardRef(
  ({ aspectRatio, children, ...props }, forwardRef) => {
    return /* @__PURE__ */ React.createElement(
      MediaPlayerBridge,
      {
        ...props,
        src: props.src,
        ref: forwardRef,
        style: {
          aspectRatio,
          ...props.style
        }
      },
      (props2) => /* @__PURE__ */ React.createElement(Primitive.div, { ...props2 }, children)
    );
  }
);
MediaPlayer.displayName = "MediaPlayer";

const MediaProviderBridge = createReactComponent(MediaProviderInstance);
const MediaProvider = React.forwardRef(
  ({ loaders = [], children, mediaProps, ...props }, forwardRef) => {
    const reactLoaders = React.useMemo(() => loaders.map((Loader) => new Loader()), loaders);
    return /* @__PURE__ */ React.createElement(MediaProviderBridge, { ...props, loaders: reactLoaders, ref: forwardRef }, (props2, instance) => /* @__PURE__ */ React.createElement("div", { ...props2 }, /* @__PURE__ */ React.createElement(MediaOutlet, { ...mediaProps, provider: instance }), children));
  }
);
MediaProvider.displayName = "MediaProvider";
function MediaOutlet({ provider, ...props }) {
  const { controls, crossOrigin, poster, remotePlaybackInfo } = useStateContext(mediaState), { loader } = provider.$state, {
    $iosControls: $$iosControls,
    $provider: $$provider,
    $providerSetup: $$providerSetup
  } = useMediaContext(), $controls = useSignal(controls), $iosControls = useSignal($$iosControls), $nativeControls = $controls || $iosControls, $crossOrigin = useSignal(crossOrigin), $poster = useSignal(poster), $loader = useSignal(loader), $provider = useSignal($$provider), $providerSetup = useSignal($$providerSetup), $remoteInfo = useSignal(remotePlaybackInfo), $mediaType = $loader?.mediaType(), isYouTubeEmbed = $loader?.name === "youtube", isVimeoEmbed = $loader?.name === "vimeo", isEmbed = isYouTubeEmbed || isVimeoEmbed, isRemotion = $loader?.name === "remotion", isGoogleCast = $loader?.name === "google-cast", [googleCastIconPaths, setGoogleCastIconPaths] = React.useState("");
  React.useEffect(() => {
    if (!isGoogleCast || googleCastIconPaths)
      return;
    import('./chunks/vidstack-DpFFzXBw.js').then((mod) => {
      setGoogleCastIconPaths(mod.default);
    });
  }, [isGoogleCast]);
  if (isGoogleCast) {
    return /* @__PURE__ */ React.createElement(
      "div",
      {
        className: "vds-google-cast",
        ref: (el) => {
          provider.load(el);
        }
      },
      /* @__PURE__ */ React.createElement(Icon, { paths: googleCastIconPaths }),
      $remoteInfo?.deviceName ? /* @__PURE__ */ React.createElement("span", { className: "vds-google-cast-info" }, "Google Cast on", " ", /* @__PURE__ */ React.createElement("span", { className: "vds-google-cast-device-name" }, $remoteInfo.deviceName)) : null
    );
  }
  if (isRemotion) {
    return /* @__PURE__ */ React.createElement("div", { "data-remotion-canvas": true }, /* @__PURE__ */ React.createElement(
      "div",
      {
        "data-remotion-container": true,
        ref: (el) => {
          provider.load(el);
        }
      },
      isRemotionProvider($provider) && $providerSetup ? React.createElement($provider.render) : null
    ));
  }
  return isEmbed ? React.createElement(
    React.Fragment,
    null,
    React.createElement("iframe", {
      className: isYouTubeEmbed ? "vds-youtube" : "vds-vimeo",
      suppressHydrationWarning: true,
      tabIndex: !$nativeControls ? -1 : void 0,
      "aria-hidden": "true",
      "data-no-controls": !$nativeControls ? "" : void 0,
      ref(el) {
        provider.load(el);
      }
    }),
    !$nativeControls ? React.createElement("div", { className: "vds-blocker" }) : null
  ) : $mediaType ? React.createElement($mediaType === "audio" ? "audio" : "video", {
    ...props,
    controls: $nativeControls ? true : null,
    crossOrigin: typeof $crossOrigin === "boolean" ? "" : $crossOrigin,
    poster: $mediaType === "video" && $nativeControls && $poster ? $poster : null,
    preload: "none",
    "aria-hidden": "true",
    suppressHydrationWarning: true,
    ref(el) {
      provider.load(el);
    }
  }) : null;
}
MediaOutlet.displayName = "MediaOutlet";

function createTextTrack(init) {
  const media = useMediaContext(), track = React.useMemo(() => new TextTrack(init), Object.values(init));
  React.useEffect(() => {
    media.textTracks.add(track);
    return () => void media.textTracks.remove(track);
  }, [track]);
  return track;
}

function Track({ lang, ...props }) {
  createTextTrack({ language: lang, ...props });
  return null;
}
Track.displayName = "Track";

const ToggleButtonBridge = createReactComponent(ToggleButtonInstance);
const ToggleButton = React.forwardRef(
  ({ children, ...props }, forwardRef) => {
    return /* @__PURE__ */ React.createElement(ToggleButtonBridge, { ...props }, (props2) => /* @__PURE__ */ React.createElement(Primitive.button, { ...props2, ref: composeRefs(props2.ref, forwardRef) }, children));
  }
);
ToggleButton.displayName = "ToggleButton";

const PosterBridge = createReactComponent(PosterInstance);
const Poster = React.forwardRef(
  ({ children, ...props }, forwardRef) => {
    return /* @__PURE__ */ React.createElement(PosterBridge, { ...props }, (props2, instance) => /* @__PURE__ */ React.createElement(PosterImg, { ...props2, instance, ref: composeRefs(props2.ref, forwardRef) }, children));
  }
);
Poster.displayName = "Poster";
const PosterImg = React.forwardRef(
  ({ instance, children, ...props }, forwardRef) => {
    const { src, img, alt, crossOrigin, loading, hidden } = instance.$state, $src = useSignal(src), $alt = useSignal(alt), $crossOrigin = useSignal(crossOrigin), $loading = useSignal(loading), $hidden = useSignal(hidden);
    return /* @__PURE__ */ React.createElement(
      Primitive.img,
      {
        ...props,
        src: $src || "",
        alt: $alt || void 0,
        crossOrigin: $crossOrigin || void 0,
        ref: composeRefs(img.set, forwardRef),
        style: { display: $loading || $hidden ? "none" : void 0 }
      },
      children
    );
  }
);
PosterImg.displayName = "PosterImg";

const Root = React.forwardRef(({ children, ...props }, forwardRef) => {
  return /* @__PURE__ */ React.createElement(
    Primitive.div,
    {
      translate: "yes",
      "aria-live": "off",
      "aria-atomic": "true",
      ...props,
      ref: forwardRef
    },
    children
  );
});
Root.displayName = "Caption";
const Text = React.forwardRef((props, forwardRef) => {
  const textTrack = useMediaState("textTrack"), [activeCue, setActiveCue] = React.useState();
  React.useEffect(() => {
    if (!textTrack)
      return;
    function onCueChange() {
      setActiveCue(textTrack?.activeCues[0]);
    }
    textTrack.addEventListener("cue-change", onCueChange);
    return () => {
      textTrack.removeEventListener("cue-change", onCueChange);
      setActiveCue(void 0);
    };
  }, [textTrack]);
  return /* @__PURE__ */ React.createElement(
    Primitive.span,
    {
      ...props,
      "data-part": "cue",
      dangerouslySetInnerHTML: {
        __html: activeCue?.text || ""
      },
      ref: forwardRef
    }
  );
});
Text.displayName = "CaptionText";

var caption = /*#__PURE__*/Object.freeze({
  __proto__: null,
  Root: Root,
  Text: Text
});

function useState(ctor, prop, ref) {
  const initialValue = React.useMemo(() => ctor.state.record[prop], [ctor, prop]);
  return useSignal(ref.current ? ref.current.$state[prop] : initialValue);
}
const storesCache = /* @__PURE__ */ new Map();
function useStore(ctor, ref) {
  const initialStore = React.useMemo(() => {
    let store = storesCache.get(ctor);
    if (!store) {
      store = new Proxy(ctor.state.record, {
        get: (_, prop) => () => ctor.state.record[prop]
      });
      storesCache.set(ctor, store);
    }
    return store;
  }, [ctor]);
  return useSignalRecord(ref.current ? ref.current.$state : initialStore);
}

function useMediaProvider() {
  const [provider, setProvider] = React.useState(null), context = useMediaContext();
  React.useEffect(() => {
    if (!context)
      return;
    return effect(() => {
      setProvider(context.$provider());
    });
  }, []);
  return provider;
}

function useThumbnails(src, crossOrigin = null) {
  useReactScope(); const $src = createSignal(src), $crossOrigin = createSignal(crossOrigin), loader = useScoped(() => ThumbnailsLoader.create($src, $crossOrigin));
  React.useEffect(() => {
    $src.set(src);
  }, [src]);
  React.useEffect(() => {
    $crossOrigin.set(crossOrigin);
  }, [crossOrigin]);
  return useSignal(loader.$images);
}
function useActiveThumbnail(thumbnails, time) {
  return React.useMemo(() => {
    let activeIndex = -1;
    for (let i = thumbnails.length - 1; i >= 0; i--) {
      const image = thumbnails[i];
      if (time >= image.startTime && (!image.endTime || time < image.endTime)) {
        activeIndex = i;
        break;
      }
    }
    return thumbnails[activeIndex] || null;
  }, [thumbnails, time]);
}

function useSliderPreview({
  clamp = false,
  offset = 0,
  orientation = "horizontal"
} = {}) {
  const [rootRef, setRootRef] = React.useState(null), [previewRef, setPreviewRef] = React.useState(null), [pointerValue, setPointerValue] = React.useState(0), [isVisible, setIsVisible] = React.useState(false);
  React.useEffect(() => {
    if (!rootRef)
      return;
    const dragging = signal(false);
    function updatePointerValue(event) {
      if (!rootRef)
        return;
      setPointerValue(getPointerValue(rootRef, event, orientation));
    }
    return effect(() => {
      if (!dragging()) {
        listenEvent(rootRef, "pointerenter", () => {
          setIsVisible(true);
          previewRef?.setAttribute("data-visible", "");
        });
        listenEvent(rootRef, "pointerdown", (event) => {
          dragging.set(true);
          updatePointerValue(event);
        });
        listenEvent(rootRef, "pointerleave", () => {
          setIsVisible(false);
          previewRef?.removeAttribute("data-visible");
        });
        listenEvent(rootRef, "pointermove", (event) => {
          updatePointerValue(event);
        });
        return;
      }
      previewRef?.setAttribute("data-dragging", "");
      listenEvent(document, "pointerup", (event) => {
        dragging.set(false);
        previewRef?.removeAttribute("data-dragging");
        updatePointerValue(event);
      });
      listenEvent(document, "pointermove", (event) => {
        updatePointerValue(event);
      });
      listenEvent(document, "touchmove", (e) => e.preventDefault(), {
        passive: false
      });
    });
  }, [rootRef]);
  React.useEffect(() => {
    if (previewRef) {
      previewRef.style.setProperty("--slider-pointer", pointerValue + "%");
    }
  }, [previewRef, pointerValue]);
  React.useEffect(() => {
    if (!previewRef)
      return;
    const update = () => {
      updateSliderPreviewPlacement(previewRef, {
        offset,
        clamp,
        orientation
      });
    };
    update();
    const resize = new ResizeObserver(update);
    resize.observe(previewRef);
    return () => resize.disconnect();
  }, [previewRef, clamp, offset, orientation]);
  return {
    previewRootRef: setRootRef,
    previewRef: setPreviewRef,
    previewValue: pointerValue,
    isPreviewVisible: isVisible
  };
}
function getPointerValue(root, event, orientation) {
  let thumbPositionRate, rect = root.getBoundingClientRect();
  if (orientation === "vertical") {
    const { bottom: trackBottom, height: trackHeight } = rect;
    thumbPositionRate = (trackBottom - event.clientY) / trackHeight;
  } else {
    const { left: trackLeft, width: trackWidth } = rect;
    thumbPositionRate = (event.clientX - trackLeft) / trackWidth;
  }
  return round(Math.max(0, Math.min(100, 100 * thumbPositionRate)));
}
function round(num) {
  return Number(num.toFixed(3));
}

export { caption as Caption, Icon, LibASSTextRenderer, Logger, MediaPlayer, MediaPlayerInstance, MediaProvider, MediaProviderInstance, Poster, PosterInstance, TextTrack, ThumbnailsLoader, ToggleButton, ToggleButtonInstance, Track, createTextTrack, mediaState, updateSliderPreviewPlacement, useActiveThumbnail, useMediaContext, useMediaProvider, useMediaState, useSliderPreview, useState, useStore, useThumbnails };
