// @vidstack/react/dist/prod/chunks/vidstack-DRk4NxIH.js
"use client"

import { h as createScope, s as signal, g as effect, p as peek, j as isString, k as deferredPromise, l as listenEvent, m as isArray } from './vidstack-BNPvxaQh.js';
import { T as TimeRange, p as preconnect, L as ListSymbol, b as TextTrack, Q as QualitySymbol } from './vidstack-BRXztYRL.js';
import { R as RAFLoop } from './vidstack-BT-_8QDd.js';
import { E as EmbedProvider, t as timedPromise } from './vidstack-Nbls93w9.js';
import { resolveVimeoVideoId, getVimeoVideoInfo } from './vidstack-BTBUzdbF.js';
import 'react';

const trackedVimeoEvents = [
  "bufferend",
  "bufferstart",
  // 'cuechange',
  "durationchange",
  "ended",
  "enterpictureinpicture",
  "error",
  "fullscreenchange",
  "leavepictureinpicture",
  "loaded",
  // 'loadeddata',
  // 'loadedmetadata',
  // 'loadstart',
  "playProgress",
  "loadProgress",
  "pause",
  "play",
  "playbackratechange",
  // 'progress',
  "qualitychange",
  "seeked",
  "seeking",
  // 'texttrackchange',
  "timeupdate",
  "volumechange",
  "waiting"
  // 'adstarted',
  // 'adcompleted',
  // 'aderror',
  // 'adskipped',
  // 'adallcompleted',
  // 'adclicked',
  // 'chapterchange',
  // 'chromecastconnected',
  // 'remoteplaybackavailabilitychange',
  // 'remoteplaybackconnecting',
  // 'remoteplaybackconnect',
  // 'remoteplaybackdisconnect',
  // 'liveeventended',
  // 'liveeventstarted',
  // 'livestreamoffline',
  // 'livestreamonline',
];

class VimeoProvider extends EmbedProvider {
  constructor(iframe, _ctx) {
    super(iframe);
    this.b = _ctx;
    this.$$PROVIDER_TYPE = "VIMEO";
    this.scope = createScope();
    this.Ga = 0;
    this.Ha = new TimeRange(0, 0);
    this.Ib = new TimeRange(0, 0);
    this.F = null;
    this.H = null;
    this.sd = null;
    this.O = signal("");
    this.pc = signal(false);
    this.td = null;
    this.W = null;
    this.fh = null;
    this.Ea = new RAFLoop(this.cd.bind(this));
    this.Xk = null;
    /**
     * Whether tracking session data should be enabled on the embed, including cookies and analytics.
     * This is turned off by default to be GDPR-compliant.
     *
     * @defaultValue `false`
     */
    this.cookies = false;
    this.title = true;
    this.byline = true;
    this.portrait = true;
    this.color = "00ADEF";
    // Embed will sometimes dispatch 0 at end of playback.
    this.Mm = false;
  }
  get c() {
    return this.b.delegate.c;
  }
  get type() {
    return "vimeo";
  }
  get currentSrc() {
    return this.W;
  }
  get videoId() {
    return this.O();
  }
  get hash() {
    return this.td;
  }
  get isPro() {
    return this.pc();
  }
  preconnect() {
    preconnect(this.fb());
  }
  setup() {
    super.setup();
    effect(this.ld.bind(this));
    effect(this.gh.bind(this));
    effect(this.hh.bind(this));
    this.c("provider-setup", this);
  }
  destroy() {
    this.I();
    this.q("destroy");
  }
  async play() {
    const { paused } = this.b.$state;
    if (!this.F) {
      this.F = timedPromise(() => {
        this.F = null;
        if (paused())
          return "Timed out.";
      });
      this.q("play");
    }
    return this.F.promise;
  }
  async pause() {
    const { paused } = this.b.$state;
    if (!this.H) {
      this.H = timedPromise(() => {
        this.H = null;
        if (!paused())
          return "Timed out.";
      });
      this.q("pause");
    }
    return this.H.promise;
  }
  setMuted(muted) {
    this.q("setMuted", muted);
  }
  setCurrentTime(time) {
    this.q("seekTo", time);
    this.c("seeking", time);
  }
  setVolume(volume) {
    this.q("setVolume", volume);
    this.q("setMuted", peek(this.b.$state.muted));
  }
  setPlaybackRate(rate) {
    this.q("setPlaybackRate", rate);
  }
  async loadSource(src) {
    if (!isString(src.src)) {
      this.W = null;
      this.td = null;
      this.O.set("");
      return;
    }
    const { videoId, hash } = resolveVimeoVideoId(src.src);
    this.O.set(videoId ?? "");
    this.td = hash ?? null;
    this.W = src;
  }
  ld() {
    this.I();
    const videoId = this.O();
    if (!videoId) {
      this.db.set("");
      return;
    }
    this.db.set(`${this.fb()}/video/${videoId}`);
    this.c("load-start");
  }
  gh() {
    const videoId = this.O();
    if (!videoId)
      return;
    const promise = deferredPromise(), abort = new AbortController();
    this.sd = promise;
    getVimeoVideoInfo(videoId, abort).then((info) => {
      promise.resolve(info);
    }).catch((e) => {
      promise.reject();
    });
    return () => {
      promise.reject();
      abort.abort();
    };
  }
  hh() {
    const isPro = this.pc(), { $state, qualities } = this.b;
    $state.canSetPlaybackRate.set(isPro);
    qualities[ListSymbol.Nc](!isPro);
    if (isPro) {
      return listenEvent(qualities, "change", () => {
        if (qualities.auto)
          return;
        const id = qualities.selected?.id;
        if (id)
          this.q("setQuality", id);
      });
    }
  }
  fb() {
    return "https://player.vimeo.com";
  }
  Ue() {
    const { $iosControls } = this.b, { keyDisabled } = this.b.$props, { controls, playsInline } = this.b.$state, showControls = controls() || $iosControls();
    return {
      title: this.title,
      byline: this.byline,
      color: this.color,
      portrait: this.portrait,
      controls: showControls,
      h: this.hash,
      keyboard: showControls && !keyDisabled(),
      transparent: true,
      playsinline: playsInline(),
      dnt: !this.cookies
    };
  }
  cd() {
    this.q("getCurrentTime");
  }
  Fb(time, trigger) {
    if (this.Mm && time === 0)
      return;
    const { realCurrentTime, realDuration, paused, bufferedEnd } = this.b.$state;
    if (realCurrentTime() === time)
      return;
    const prevTime = realCurrentTime(), detail = {
      currentTime: time,
      played: this.Ek(time)
    };
    this.c("time-update", detail, trigger);
    if (Math.abs(prevTime - time) > 1.5) {
      this.c("seeking", time, trigger);
      if (!paused() && bufferedEnd() < time) {
        this.c("waiting", void 0, trigger);
      }
    }
    if (realDuration() - time < 0.01) {
      this.c("end", void 0, trigger);
      this.Mm = true;
      setTimeout(() => {
        this.Mm = false;
      }, 500);
    }
  }
  Ek(time) {
    return this.Ga >= time ? this.Ha : this.Ha = new TimeRange(0, this.Ga = time);
  }
  cb(time, trigger) {
    this.c("seeked", time, trigger);
  }
  rd(trigger) {
    const videoId = this.O();
    this.sd?.promise.then((info) => {
      if (!info)
        return;
      const { title, poster, duration, pro } = info;
      this.pc.set(pro);
      this.c("title-change", title, trigger);
      this.c("poster-change", poster, trigger);
      this.c("duration-change", duration, trigger);
      this.nd(duration, trigger);
    }).catch(() => {
      if (videoId !== this.O())
        return;
      this.q("getVideoTitle");
      this.q("getDuration");
    });
  }
  nd(duration, trigger) {
    const { $iosControls } = this.b, { controls } = this.b.$state, showEmbedControls = controls() || $iosControls();
    this.Ib = new TimeRange(0, duration);
    const detail = {
      buffered: new TimeRange(0, 0),
      seekable: this.Ib,
      duration
    };
    this.b.delegate.kc(detail, trigger);
    if (!showEmbedControls) {
      this.q("_hideOverlay");
    }
    this.q("getQualities");
    this.q("getChapters");
  }
  ih(method, data, trigger) {
    switch (method) {
      case "getVideoTitle":
        const videoTitle = data;
        this.c("title-change", videoTitle, trigger);
        break;
      case "getDuration":
        const duration = data;
        if (!this.b.$state.canPlay()) {
          this.nd(duration, trigger);
        } else {
          this.c("duration-change", duration, trigger);
        }
        break;
      case "getCurrentTime":
        this.Fb(data, trigger);
        break;
      case "getBuffered":
        if (isArray(data) && data.length) {
          this.Ze(data[data.length - 1][1], trigger);
        }
        break;
      case "setMuted":
        this.bb(peek(this.b.$state.volume), data, trigger);
        break;
      case "getChapters":
        this.ol(data);
        break;
      case "getQualities":
        this.qc(data, trigger);
        break;
    }
  }
  jh() {
    for (const type of trackedVimeoEvents) {
      this.q("addEventListener", type);
    }
  }
  Ba(trigger) {
    this.Ea.sa();
    this.c("pause", void 0, trigger);
    this.H?.resolve();
    this.H = null;
  }
  yb(trigger) {
    this.Ea.Cb();
    this.c("play", void 0, trigger);
    this.F?.resolve();
    this.F = null;
  }
  kh(trigger) {
    const { paused } = this.b.$state;
    if (!paused()) {
      this.c("playing", void 0, trigger);
    }
  }
  Ze(buffered, trigger) {
    const detail = {
      buffered: new TimeRange(0, buffered),
      seekable: this.Ib
    };
    this.c("progress", detail, trigger);
  }
  lh(trigger) {
    this.c("waiting", void 0, trigger);
  }
  mh(trigger) {
    const { paused } = this.b.$state;
    if (!paused())
      this.c("playing", void 0, trigger);
  }
  ed(trigger) {
    const { paused } = this.b.$state;
    if (paused()) {
      this.c("play", void 0, trigger);
    }
    this.c("waiting", void 0, trigger);
  }
  bb(volume, muted, trigger) {
    const detail = { volume, muted };
    this.c("volume-change", detail, trigger);
  }
  // protected _onTextTrackChange(track: VimeoTextTrack, trigger: Event) {
  //   const textTrack = this._ctx.textTracks.toArray().find((t) => t.language === track.language);
  //   if (textTrack) textTrack.mode = track.mode;
  // }
  // protected _onTextTracksChange(tracks: VimeoTextTrack[], trigger: Event) {
  //   for (const init of tracks) {
  //     const textTrack = new TextTrack({
  //       ...init,
  //       label: init.label.replace('auto-generated', 'auto'),
  //     });
  //     textTrack[TextTrackSymbol._readyState] = 2;
  //     this._ctx.textTracks.add(textTrack, trigger);
  //     textTrack.setMode(init.mode, trigger);
  //   }
  // }
  // protected _onCueChange(cue: VimeoTextCue, trigger: Event) {
  //   const { textTracks, $state } = this._ctx,
  //     { currentTime } = $state,
  //     track = textTracks.selected;
  //   if (this._currentCue) track?.removeCue(this._currentCue, trigger);
  //   this._currentCue = new window.VTTCue(currentTime(), Number.MAX_SAFE_INTEGER, cue.text);
  //   track?.addCue(this._currentCue, trigger);
  // }
  ol(chapters) {
    this.cl();
    if (!chapters.length)
      return;
    const track = new TextTrack({
      kind: "chapters",
      default: true
    }), { realDuration } = this.b.$state;
    for (let i = 0; i < chapters.length; i++) {
      const chapter = chapters[i], nextChapter = chapters[i + 1];
      track.addCue(
        new window.VTTCue(
          chapter.startTime,
          nextChapter?.startTime ?? realDuration(),
          chapter.title
        )
      );
    }
    this.Xk = track;
    this.b.textTracks.add(track);
  }
  cl() {
    if (!this.Xk)
      return;
    this.b.textTracks.remove(this.Xk);
    this.Xk = null;
  }
  qc(qualities, trigger) {
    this.b.qualities[QualitySymbol._a] = qualities.some((q) => q.id === "auto") ? () => {
      this.q("setQuality", "auto");
    } : void 0;
    for (const quality of qualities) {
      if (quality.id === "auto")
        continue;
      const height = +quality.id.slice(0, -1);
      if (isNaN(height))
        continue;
      this.b.qualities[ListSymbol.pa](
        {
          id: quality.id,
          width: height * (16 / 9),
          height,
          codec: "avc1,h.264",
          bitrate: -1
        },
        trigger
      );
    }
    this.gb(
      qualities.find((q) => q.active),
      trigger
    );
  }
  gb({ id } = {}, trigger) {
    if (!id)
      return;
    const isAuto = id === "auto", newQuality = this.b.qualities.toArray().find((q) => q.id === id);
    if (isAuto) {
      this.b.qualities[QualitySymbol.Za](isAuto, trigger);
      this.b.qualities[ListSymbol.qa](void 0, true, trigger);
    } else {
      this.b.qualities[ListSymbol.qa](newQuality, true, trigger);
    }
  }
  nh(event, payload, trigger) {
    switch (event) {
      case "ready":
        this.jh();
        break;
      case "loaded":
        this.rd(trigger);
        break;
      case "play":
        this.yb(trigger);
        break;
      case "playProgress":
        this.kh(trigger);
        break;
      case "pause":
        this.Ba(trigger);
        break;
      case "loadProgress":
        this.Ze(payload.seconds, trigger);
        break;
      case "waiting":
        this.ed(trigger);
        break;
      case "bufferstart":
        this.lh(trigger);
        break;
      case "bufferend":
        this.mh(trigger);
        break;
      case "volumechange":
        this.bb(payload.volume, peek(this.b.$state.muted), trigger);
        break;
      case "durationchange":
        this.Ib = new TimeRange(0, payload.duration);
        this.c("duration-change", payload.duration, trigger);
        break;
      case "playbackratechange":
        this.c("rate-change", payload.playbackRate, trigger);
        break;
      case "qualitychange":
        this.gb(payload, trigger);
        break;
      case "fullscreenchange":
        this.c("fullscreen-change", payload.fullscreen, trigger);
        break;
      case "enterpictureinpicture":
        this.c("picture-in-picture-change", true, trigger);
        break;
      case "leavepictureinpicture":
        this.c("picture-in-picture-change", false, trigger);
        break;
      case "ended":
        this.c("end", void 0, trigger);
        break;
      case "error":
        this.V(payload, trigger);
        break;
      case "seek":
      case "seeked":
        this.cb(payload.seconds, trigger);
        break;
    }
  }
  V(error, trigger) {
    if (error.method === "setPlaybackRate") {
      this.pc.set(false);
    }
    if (error.method === "play") {
      this.F?.reject(error.message);
      return;
    }
  }
  jd(message, event) {
    if (message.event) {
      this.nh(message.event, message.data, event);
    } else if (message.method) {
      this.ih(message.method, message.value, event);
    }
  }
  mc() {
  }
  q(command, arg) {
    return this.hd({
      method: command,
      value: arg
    });
  }
  I() {
    this.Ea.sa();
    this.Ga = 0;
    this.Ha = new TimeRange(0, 0);
    this.Ib = new TimeRange(0, 0);
    this.F = null;
    this.H = null;
    this.sd = null;
    this.fh = null;
    this.pc.set(false);
    this.cl();
  }
}

export { VimeoProvider };
