// @vidstack/react/dist/prod/chunks/vidstack-Nbls93w9.js
"use client"

import { k as deferredPromise, s as signal, l as listenEvent, g as effect, p as peek, j as isString } from './vidstack-BNPvxaQh.js';
import { c as appendParamsToURL, I as IS_SERVER } from './vidstack-BRXztYRL.js';

function timedPromise(callback, ms = 3e3) {
  const promise = deferredPromise();
  setTimeout(() => {
    const rejection = callback();
    if (rejection)
      promise.reject(rejection);
  }, ms);
  return promise;
}

class EmbedProvider {
  constructor(_iframe) {
    this.eb = _iframe;
    this.db = signal("");
    /**
     * Defines which referrer is sent when fetching the resource.
     *
     * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLIFrameElement/referrerPolicy}
     */
    this.referrerPolicy = null;
    _iframe.setAttribute("frameBorder", "0");
    _iframe.setAttribute("aria-hidden", "true");
    _iframe.setAttribute(
      "allow",
      "autoplay; fullscreen; encrypted-media; picture-in-picture; accelerometer; gyroscope"
    );
    if (this.referrerPolicy !== null) {
      _iframe.setAttribute("referrerpolicy", this.referrerPolicy);
    }
  }
  get iframe() {
    return this.eb;
  }
  setup() {
    listenEvent(window, "message", this.Mg.bind(this));
    listenEvent(this.eb, "load", this.mc.bind(this));
    effect(this.gd.bind(this));
  }
  gd() {
    const src = this.db();
    if (!src.length) {
      this.eb.setAttribute("src", "");
      return;
    }
    const params = peek(() => this.Ue());
    this.eb.setAttribute("src", appendParamsToURL(src, params));
  }
  hd(message, target) {
    if (IS_SERVER)
      return;
    this.eb.contentWindow?.postMessage(JSON.stringify(message), target ?? "*");
  }
  Mg(event) {
    const origin = this.fb(), isOriginMatch = (event.source === null || event.source === this.eb?.contentWindow) && (!isString(origin) || origin === event.origin);
    if (!isOriginMatch)
      return;
    try {
      const message = JSON.parse(event.data);
      if (message)
        this.jd(message, event);
      return;
    } catch (e) {
    }
    if (event.data)
      this.jd(event.data, event);
  }
}

export { EmbedProvider as E, timedPromise as t };
