// @vidstack/react/dist/prod/chunks/vidstack-DShZT0NX.js
"use client"

import { l as listen, d as IS_CHROME, e as IS_IOS, f as canGoogleCastSrc, g as coerceToError, I as IS_SERVER, h as loadScript } from './vidstack-BRXztYRL.js';
import { p as peek, D as DOMEvent } from './vidstack-BNPvxaQh.js';

function getCastFrameworkURL() {
  return "https://www.gstatic.com/cv/js/sender/v1/cast_sender.js?loadCastFramework=1";
}
function hasLoadedCastFramework() {
  return !!window.cast?.framework;
}
function isCastAvailable() {
  return !!window.chrome?.cast?.isAvailable;
}
function isCastConnected() {
  return getCastContext().getCastState() === cast.framework.CastState.CONNECTED;
}
function getCastContext() {
  return window.cast.framework.CastContext.getInstance();
}
function getCastSession() {
  return getCastContext().getCurrentSession();
}
function getCastSessionMedia() {
  return getCastSession()?.getSessionObj().media[0];
}
function hasActiveCastSession(src) {
  const contentId = getCastSessionMedia()?.media.contentId;
  return contentId === src?.src;
}
function getDefaultCastOptions() {
  return {
    language: "en-US",
    autoJoinPolicy: chrome.cast.AutoJoinPolicy.ORIGIN_SCOPED,
    receiverApplicationId: chrome.cast.media.DEFAULT_MEDIA_RECEIVER_APP_ID,
    resumeSavedSession: true,
    androidReceiverCompatible: true
  };
}
function getCastErrorMessage(code) {
  const defaultMessage = `Google Cast Error Code: ${code}`;
  return defaultMessage;
}
function listenCastContextEvent(type, handler) {
  return listen(getCastContext(), type, handler);
}

class GoogleCastLoader {
  constructor() {
    this.name = "google-cast";
  }
  /**
   * @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.CastContext}
   */
  get cast() {
    return getCastContext();
  }
  mediaType() {
    return "video";
  }
  canPlay(src) {
    return IS_CHROME && !IS_IOS && canGoogleCastSrc(src);
  }
  async prompt(ctx) {
    let openEvent;
    try {
      const loadEvent = await this.wm(ctx);
      if (!this.ba) {
        this.ba = new cast.framework.RemotePlayer();
        new cast.framework.RemotePlayerController(this.ba);
      }
      openEvent = ctx.player.createEvent("google-cast-prompt-open", {
        trigger: loadEvent
      });
      ctx.player.dispatchEvent(openEvent);
      this.Nl(ctx, "connecting", openEvent);
      await this.xm(peek(ctx.$props.googleCast));
      ctx.$state.remotePlaybackInfo.set({
        deviceName: getCastSession()?.getCastDevice().friendlyName
      });
      if (isCastConnected())
        this.Nl(ctx, "connected", openEvent);
    } catch (message) {
      const error = coerceToError(message);
      this.Nl(
        ctx,
        isCastConnected() ? "connected" : "disconnected",
        new DOMEvent("google-cast-prompt-error", { detail: error })
      );
      throw error;
    } finally {
      ctx.player.dispatch("google-cast-prompt-close", {
        trigger: openEvent
      });
    }
  }
  async load(ctx) {
    if (IS_SERVER) {
      throw Error("[vidstack] can not load google cast provider server-side");
    }
    if (!this.ba) {
      throw Error("[vidstack] google cast player was not initialized");
    }
    return new (await import('./vidstack-Cxyb4r5N.js')).GoogleCastProvider(this.ba, ctx);
  }
  async wm(ctx) {
    if (hasLoadedCastFramework())
      return;
    const loadStartEvent = ctx.player.createEvent("google-cast-load-start");
    ctx.player.dispatch(loadStartEvent);
    await loadScript(getCastFrameworkURL());
    await customElements.whenDefined("google-cast-launcher");
    const loadedEvent = ctx.player.createEvent("google-cast-loaded", { trigger: loadStartEvent });
    ctx.player.dispatch(loadedEvent);
    if (!isCastAvailable()) {
      throw Error(
        "Cast not available."
      );
    }
    return loadedEvent;
  }
  async xm(options) {
    this.ym(options);
    const errorCode = await this.cast.requestSession();
    if (errorCode)
      throw Error(getCastErrorMessage(errorCode));
  }
  ym(options) {
    this.cast?.setOptions({
      ...getDefaultCastOptions(),
      ...options
    });
  }
  Nl(ctx, state, trigger) {
    const detail = { type: "google-cast", state };
    ctx.delegate.c("remote-playback-change", detail, trigger);
  }
}

var loader = /*#__PURE__*/Object.freeze({
  __proto__: null,
  GoogleCastLoader: GoogleCastLoader
});

export { getCastContext as a, getCastSession as b, getCastErrorMessage as c, loader as d, getCastSessionMedia as g, hasActiveCastSession as h, listenCastContextEvent as l };
