// @vidstack/react/dist/prod/chunks/vidstack-Cxyb4r5N.js
"use client"

import { l as listenEvent, g as effect, z as untrack, h as createScope, A as keysOf, w as onDispose, D as DOMEvent, p as peek } from './vidstack-BNPvxaQh.js';
import { L as ListSymbol, T as TimeRange } from './vidstack-BRXztYRL.js';
import { R as RAFLoop } from './vidstack-BT-_8QDd.js';
import { g as getCastSessionMedia, a as getCastContext, b as getCastSession, h as hasActiveCastSession, l as listenCastContextEvent, c as getCastErrorMessage } from './vidstack-DShZT0NX.js';
import 'react';

class GoogleCastMediaInfoBuilder {
  constructor(src) {
    this.El = new chrome.cast.media.MediaInfo(src.src, src.type);
  }
  build() {
    return this.El;
  }
  dm(streamType) {
    if (streamType.includes("live")) {
      this.El.streamType = chrome.cast.media.StreamType.LIVE;
    } else {
      this.El.streamType = chrome.cast.media.StreamType.BUFFERED;
    }
    return this;
  }
  em(tracks) {
    this.El.tracks = tracks.map(this.fm);
    return this;
  }
  gm(title, poster) {
    this.El.metadata = new chrome.cast.media.GenericMediaMetadata();
    this.El.metadata.title = title;
    this.El.metadata.images = [{ url: poster }];
    return this;
  }
  fm(track, trackId) {
    const castTrack = new chrome.cast.media.Track(trackId, chrome.cast.media.TrackType.TEXT);
    castTrack.name = track.label;
    castTrack.trackContentId = track.src;
    castTrack.trackContentType = "text/vtt";
    castTrack.language = track.language;
    castTrack.subtype = track.kind.toUpperCase();
    return castTrack;
  }
}

const REMOTE_TRACK_TEXT_TYPE = chrome.cast.media.TrackType.TEXT, REMOTE_TRACK_AUDIO_TYPE = chrome.cast.media.TrackType.AUDIO;
class GoogleCastTracksManager {
  constructor(_cast, _ctx, _onNewLocalTracks) {
    this.Il = _cast;
    this.b = _ctx;
    this.Ll = _onNewLocalTracks;
  }
  mk() {
    const syncRemoteActiveIds = this.Ql.bind(this);
    listenEvent(this.b.audioTracks, "change", syncRemoteActiveIds);
    listenEvent(this.b.textTracks, "mode-change", syncRemoteActiveIds);
    effect(this.hm.bind(this));
  }
  Hl() {
    return this.b.$state.textTracks().filter((track) => track.src && track.type === "vtt");
  }
  Rl() {
    return this.b.$state.audioTracks();
  }
  Gl(type) {
    const tracks = this.Il.mediaInfo?.tracks ?? [];
    return type ? tracks.filter((track) => track.type === type) : tracks;
  }
  im() {
    const activeIds = [], activeLocalAudioTrack = this.Rl().find((track) => track.selected), activeLocalTextTracks = this.Hl().filter(
      (track) => track.mode === "showing"
    );
    if (activeLocalAudioTrack) {
      const remoteAudioTracks = this.Gl(REMOTE_TRACK_AUDIO_TYPE), remoteAudioTrack = this.Kl(remoteAudioTracks, activeLocalAudioTrack);
      if (remoteAudioTrack)
        activeIds.push(remoteAudioTrack.trackId);
    }
    if (activeLocalTextTracks?.length) {
      const remoteTextTracks = this.Gl(REMOTE_TRACK_TEXT_TYPE);
      if (remoteTextTracks.length) {
        for (const localTrack of activeLocalTextTracks) {
          const remoteTextTrack = this.Kl(remoteTextTracks, localTrack);
          if (remoteTextTrack)
            activeIds.push(remoteTextTrack.trackId);
        }
      }
    }
    return activeIds;
  }
  hm() {
    const localTextTracks = this.Hl();
    if (!this.Il.isMediaLoaded)
      return;
    const remoteTextTracks = this.Gl(REMOTE_TRACK_TEXT_TYPE);
    for (const localTrack of localTextTracks) {
      const hasRemoteTrack = this.Kl(remoteTextTracks, localTrack);
      if (!hasRemoteTrack) {
        untrack(() => this.Ll?.());
        break;
      }
    }
  }
  jm(event) {
    if (!this.Il.isMediaLoaded)
      return;
    const localAudioTracks = this.Rl(), localTextTracks = this.Hl(), remoteAudioTracks = this.Gl(REMOTE_TRACK_AUDIO_TYPE), remoteTextTracks = this.Gl(REMOTE_TRACK_TEXT_TYPE);
    for (const remoteAudioTrack of remoteAudioTracks) {
      const hasLocalTrack = this.Sl(localAudioTracks, remoteAudioTrack);
      if (hasLocalTrack)
        continue;
      const localAudioTrack = {
        id: remoteAudioTrack.trackId.toString(),
        label: remoteAudioTrack.name,
        language: remoteAudioTrack.language,
        kind: remoteAudioTrack.subtype ?? "main",
        selected: false
      };
      this.b.audioTracks[ListSymbol.pa](localAudioTrack, event);
    }
    for (const remoteTextTrack of remoteTextTracks) {
      const hasLocalTrack = this.Sl(localTextTracks, remoteTextTrack);
      if (hasLocalTrack)
        continue;
      const localTextTrack = {
        id: remoteTextTrack.trackId.toString(),
        src: remoteTextTrack.trackContentId,
        label: remoteTextTrack.name,
        language: remoteTextTrack.language,
        kind: remoteTextTrack.subtype.toLowerCase()
      };
      this.b.textTracks.add(localTextTrack, event);
    }
  }
  Ql(event) {
    if (!this.Il.isMediaLoaded)
      return;
    const activeIds = this.im(), editRequest = new chrome.cast.media.EditTracksInfoRequest(activeIds);
    this.km(editRequest).catch((error) => {
    });
  }
  km(request) {
    const media = getCastSessionMedia();
    return new Promise((resolve, reject) => media?.editTracksInfo(request, resolve, reject));
  }
  Sl(localTracks, remoteTrack) {
    return localTracks.find((localTrack) => this.Tl(localTrack, remoteTrack));
  }
  Kl(remoteTracks, localTrack) {
    return remoteTracks.find((remoteTrack) => this.Tl(localTrack, remoteTrack));
  }
  // Note: we can't rely on id matching because they will differ between local/remote. A local
  // track id might not even exist.
  Tl(localTrack, remoteTrack) {
    return remoteTrack.name === localTrack.label && remoteTrack.language === localTrack.language && remoteTrack.subtype.toLowerCase() === localTrack.kind.toLowerCase();
  }
}

class GoogleCastProvider {
  constructor(_player, _ctx) {
    this.ba = _player;
    this.b = _ctx;
    this.$$PROVIDER_TYPE = "GOOGLE_CAST";
    this.scope = createScope();
    this.W = null;
    this.nc = "disconnected";
    this.Wa = 0;
    this.Ga = 0;
    this.Ha = new TimeRange(0, 0);
    this.Ib = new TimeRange(0, 0);
    this.Ea = new RAFLoop(this.cd.bind(this));
    this.Cl = null;
    this.Ml = false;
    this.zb = new GoogleCastTracksManager(
      this.ba,
      this.b,
      this.Ll.bind(this)
    );
  }
  get c() {
    return this.b.delegate.c;
  }
  get type() {
    return "google-cast";
  }
  get currentSrc() {
    return this.W;
  }
  /**
   * The Google Cast remote player.
   *
   * @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.RemotePlayer}
   */
  get player() {
    return this.ba;
  }
  /**
   * @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.CastContext}
   */
  get cast() {
    return getCastContext();
  }
  /**
   * @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.CastSession}
   */
  get session() {
    return getCastSession();
  }
  /**
   * @see {@link https://developers.google.com/cast/docs/reference/web_sender/chrome.cast.media.Media}
   */
  get media() {
    return getCastSessionMedia();
  }
  /**
   * Whether the current Google Cast session belongs to this provider.
   */
  get hasActiveSession() {
    return hasActiveCastSession(this.W);
  }
  setup() {
    this.lm();
    this.mm();
    this.zb.mk();
    this.c("provider-setup", this);
  }
  lm() {
    listenCastContextEvent(
      cast.framework.CastContextEventType.CAST_STATE_CHANGED,
      this.Vl.bind(this)
    );
  }
  mm() {
    const Event2 = cast.framework.RemotePlayerEventType, handlers = {
      [Event2.IS_CONNECTED_CHANGED]: this.Vl,
      [Event2.IS_MEDIA_LOADED_CHANGED]: this.Wl,
      [Event2.CAN_CONTROL_VOLUME_CHANGED]: this.Xl,
      [Event2.CAN_SEEK_CHANGED]: this.Yl,
      [Event2.DURATION_CHANGED]: this.xg,
      [Event2.IS_MUTED_CHANGED]: this.bb,
      [Event2.VOLUME_LEVEL_CHANGED]: this.bb,
      [Event2.IS_PAUSED_CHANGED]: this.nm,
      [Event2.LIVE_SEEKABLE_RANGE_CHANGED]: this.jc,
      [Event2.PLAYER_STATE_CHANGED]: this.om
    };
    this.Ul = handlers;
    const handler = this.pm.bind(this);
    for (const type of keysOf(handlers)) {
      this.ba.controller.addEventListener(type, handler);
    }
    onDispose(() => {
      for (const type of keysOf(handlers)) {
        this.ba.controller.removeEventListener(type, handler);
      }
    });
  }
  async play() {
    if (!this.ba.isPaused && !this.Ml)
      return;
    if (this.Ml) {
      await this.Zl(false, 0);
      return;
    }
    this.ba.controller?.playOrPause();
  }
  async pause() {
    if (this.ba.isPaused)
      return;
    this.ba.controller?.playOrPause();
  }
  getMediaStatus(request) {
    return new Promise((resolve, reject) => {
      this.media?.getStatus(request, resolve, reject);
    });
  }
  setMuted(muted) {
    const hasChanged = muted && !this.ba.isMuted || !muted && this.ba.isMuted;
    if (hasChanged)
      this.ba.controller?.muteOrUnmute();
  }
  setCurrentTime(time) {
    this.ba.currentTime = time;
    this.c("seeking", time);
    this.ba.controller?.seek();
  }
  setVolume(volume) {
    this.ba.volumeLevel = volume;
    this.ba.controller?.setVolumeLevel();
  }
  async loadSource(src) {
    if (this.Cl?.src !== src)
      this.Cl = null;
    if (hasActiveCastSession(src)) {
      this.qm();
      this.W = src;
      return;
    }
    this.c("load-start");
    const loadRequest = this.rm(src), errorCode = await this.session.loadMedia(loadRequest);
    if (errorCode) {
      this.W = null;
      this.c("error", Error(getCastErrorMessage(errorCode)));
      return;
    }
    this.W = src;
  }
  destroy() {
    this.I();
    this._l();
  }
  I() {
    if (!this.Cl) {
      this.Ga = 0;
      this.Ha = new TimeRange(0, 0);
      this.Ib = new TimeRange(0, 0);
    }
    this.Ea.sa();
    this.Wa = 0;
    this.Cl = null;
  }
  qm() {
    const resumeSessionEvent = new DOMEvent("resume-session", { detail: this.session });
    this.Wl(resumeSessionEvent);
    const { muted, volume, remotePlaybackInfo } = this.b.$state, localState = remotePlaybackInfo();
    this.setCurrentTime(
      Math.max(this.ba.currentTime, localState?.savedState?.currentTime ?? 0)
    );
    this.setMuted(muted());
    this.setVolume(volume());
    if (localState?.savedState?.paused === false)
      this.play();
  }
  _l() {
    this.cast.endCurrentSession(true);
    const { remotePlaybackLoader } = this.b.$state;
    remotePlaybackLoader.set(null);
  }
  sm() {
    this.b.$state.remotePlaybackInfo.set({
      savedState: {
        paused: this.ba.isPaused,
        currentTime: this.ba.currentTime
      }
    });
    this._l();
  }
  cd() {
    this.tm();
  }
  pm(event) {
    this.Ul[event.type].call(this, event);
  }
  Vl(data) {
    const castState = this.cast.getCastState(), state = castState === cast.framework.CastState.CONNECTED ? "connected" : castState === cast.framework.CastState.CONNECTING ? "connecting" : "disconnected";
    if (this.nc === state)
      return;
    const detail = { type: "google-cast", state }, trigger = this.Fl(data);
    this.nc = state;
    this.c("remote-playback-change", detail, trigger);
    if (state === "disconnected") {
      this.sm();
    }
  }
  Wl(event) {
    const hasLoaded = !!this.ba.isMediaLoaded;
    if (!hasLoaded)
      return;
    const src = peek(this.b.$state.source);
    Promise.resolve().then(() => {
      if (src !== peek(this.b.$state.source) || !this.ba.isMediaLoaded)
        return;
      this.I();
      const duration = this.ba.duration;
      this.Ib = new TimeRange(0, duration);
      const detail = {
        provider: this,
        duration,
        buffered: this.Ha,
        seekable: this.$l()
      }, trigger = this.Fl(event);
      this.c("loaded-metadata", void 0, trigger);
      this.c("loaded-data", void 0, trigger);
      this.c("can-play", detail, trigger);
      this.Xl();
      this.Yl(event);
      const { volume, muted } = this.b.$state;
      this.setVolume(volume());
      this.setMuted(muted());
      this.Ea.Cb();
      this.zb.jm(trigger);
      this.zb.Ql(trigger);
    });
  }
  Xl() {
    this.b.$state.canSetVolume.set(this.ba.canControlVolume);
  }
  Yl(event) {
    const trigger = this.Fl(event);
    this.c("stream-type-change", this.um(), trigger);
  }
  um() {
    const streamType = this.ba.mediaInfo?.streamType;
    return streamType === chrome.cast.media.StreamType.LIVE ? this.ba.canSeek ? "live:dvr" : "live" : "on-demand";
  }
  tm() {
    if (this.Cl)
      return;
    const currentTime = this.ba.currentTime;
    if (currentTime === this.Wa)
      return;
    const prevPlayed = this.Ga, played = this.Ek(currentTime), detail = { currentTime, played };
    this.c("time-update", detail);
    if (currentTime > prevPlayed)
      this.jc();
    if (this.b.$state.seeking()) {
      this.c("seeked", currentTime);
    }
    this.Wa = currentTime;
  }
  Ek(time) {
    return this.Ga >= time ? this.Ha : this.Ha = new TimeRange(0, this.Ga = time);
  }
  xg(event) {
    if (!this.ba.isMediaLoaded || this.Cl)
      return;
    const duration = this.ba.duration, trigger = this.Fl(event);
    this.Ib = new TimeRange(0, duration);
    this.c("duration-change", duration, trigger);
  }
  bb(event) {
    if (!this.ba.isMediaLoaded)
      return;
    const detail = {
      muted: this.ba.isMuted,
      volume: this.ba.volumeLevel
    }, trigger = this.Fl(event);
    this.c("volume-change", detail, trigger);
  }
  nm(event) {
    const trigger = this.Fl(event);
    if (this.ba.isPaused) {
      this.c("pause", void 0, trigger);
    } else {
      this.c("play", void 0, trigger);
    }
  }
  jc(event) {
    const detail = {
      seekable: this.$l(),
      buffered: this.Ha
    }, trigger = event ? this.Fl(event) : void 0;
    this.c("progress", detail, trigger);
  }
  om(event) {
    const state = this.ba.playerState, PlayerState = chrome.cast.media.PlayerState;
    this.Ml = state === PlayerState.IDLE;
    if (state === PlayerState.PAUSED)
      return;
    const trigger = this.Fl(event);
    switch (state) {
      case PlayerState.PLAYING:
        this.c("playing", void 0, trigger);
        break;
      case PlayerState.BUFFERING:
        this.c("waiting", void 0, trigger);
        break;
      case PlayerState.IDLE:
        this.Ea.sa();
        this.c("pause");
        this.c("end");
        break;
    }
  }
  $l() {
    return this.ba.liveSeekableRange ? new TimeRange(this.ba.liveSeekableRange.start, this.ba.liveSeekableRange.end) : this.Ib;
  }
  Fl(detail) {
    return detail instanceof Event ? detail : new DOMEvent(detail.type, { detail });
  }
  vm(src) {
    const { streamType, title, poster } = this.b.$state;
    return new GoogleCastMediaInfoBuilder(src).gm(title(), poster()).dm(streamType()).em(this.zb.Hl()).build();
  }
  rm(src) {
    const mediaInfo = this.vm(src), request = new chrome.cast.media.LoadRequest(mediaInfo), info = this.b.$state.remotePlaybackInfo();
    request.autoplay = (this.Cl?.paused ?? info?.savedState?.paused) === false;
    request.currentTime = this.Cl?.time ?? info?.savedState?.currentTime ?? 0;
    return request;
  }
  async Zl(paused, time) {
    const src = peek(this.b.$state.source);
    this.Cl = { src, paused, time };
    await this.loadSource(src);
  }
  Ll() {
    this.Zl(this.ba.isPaused, this.ba.currentTime).catch((error) => {
    });
  }
}

export { GoogleCastProvider };
