// @vidstack/react/dist/prod/chunks/vidstack-BDEG6RNK.js
"use client"

import { h as createScope, s as signal, g as effect, j as isString, n as isObject, c as isNumber, o as isBoolean } from './vidstack-BNPvxaQh.js';
import { T as TimeRange, p as preconnect } from './vidstack-BRXztYRL.js';
import { E as EmbedProvider, t as timedPromise } from './vidstack-Nbls93w9.js';
import { resolveYouTubeVideoId } from './vidstack-DscYSLiW.js';
import 'react';

const YouTubePlayerState = {
  Rj: -1,
  Ve: 0,
  We: 1,
  Ng: 2,
  Og: 3,
  Pg: 5
};

class YouTubeProvider extends EmbedProvider {
  constructor(iframe, _ctx) {
    super(iframe);
    this.b = _ctx;
    this.$$PROVIDER_TYPE = "YOUTUBE";
    this.scope = createScope();
    this.O = signal("");
    this.nc = -1;
    this.oc = -1;
    this.Wk = false;
    this.Ga = 0;
    this.Ha = new TimeRange(0, 0);
    this.W = null;
    this.F = null;
    this.H = null;
    /**
     * Sets the player's interface language. The parameter value is an ISO 639-1 two-letter
     * language code or a fully specified locale. For example, fr and fr-ca are both valid values.
     * Other language input codes, such as IETF language tags (BCP 47) might also be handled properly.
     *
     * The interface language is used for tooltips in the player and also affects the default caption
     * track. Note that YouTube might select a different caption track language for a particular
     * user based on the user's individual language preferences and the availability of caption tracks.
     *
     * @defaultValue 'en'
     */
    this.language = "en";
    this.color = "red";
    /**
     * Whether cookies should be enabled on the embed. This is turned off by default to be
     * GDPR-compliant.
     *
     * @defaultValue `false`
     */
    this.cookies = false;
  }
  get c() {
    return this.b.delegate.c;
  }
  get currentSrc() {
    return this.W;
  }
  get type() {
    return "youtube";
  }
  get videoId() {
    return this.O();
  }
  preconnect() {
    preconnect(this.fb());
  }
  setup() {
    super.setup();
    effect(this.ld.bind(this));
    this.c("provider-setup", this);
  }
  async play() {
    const { paused } = this.b.$state;
    if (!this.F) {
      this.F = timedPromise(() => {
        this.F = null;
        if (paused())
          return "Timed out.";
      });
      this.q("playVideo");
    }
    return this.F.promise;
  }
  async pause() {
    const { paused } = this.b.$state;
    if (!this.H) {
      this.H = timedPromise(() => {
        this.H = null;
        if (!paused())
          ;
      });
      this.q("pauseVideo");
    }
    return this.H.promise;
  }
  setMuted(muted) {
    if (muted)
      this.q("mute");
    else
      this.q("unMute");
  }
  setCurrentTime(time) {
    this.Wk = this.b.$state.paused();
    this.q("seekTo", time);
    this.c("seeking", time);
  }
  setVolume(volume) {
    this.q("setVolume", volume * 100);
  }
  setPlaybackRate(rate) {
    this.q("setPlaybackRate", rate);
  }
  async loadSource(src) {
    if (!isString(src.src)) {
      this.W = null;
      this.O.set("");
      return;
    }
    const videoId = resolveYouTubeVideoId(src.src);
    this.O.set(videoId ?? "");
    this.W = src;
  }
  fb() {
    return !this.cookies ? "https://www.youtube-nocookie.com" : "https://www.youtube.com";
  }
  ld() {
    this.I();
    const videoId = this.O();
    if (!videoId) {
      this.db.set("");
      return;
    }
    this.db.set(`${this.fb()}/embed/${videoId}`);
    this.c("load-start");
  }
  Ue() {
    const { keyDisabled } = this.b.$props, { $iosControls } = this.b, { controls, muted, playsInline } = this.b.$state, showControls = controls() || $iosControls();
    return {
      autoplay: 0,
      cc_lang_pref: this.language,
      cc_load_policy: showControls ? 1 : void 0,
      color: this.color,
      controls: showControls ? 1 : 0,
      disablekb: !showControls || keyDisabled() ? 1 : 0,
      enablejsapi: 1,
      fs: 1,
      hl: this.language,
      iv_load_policy: showControls ? 1 : 3,
      mute: muted() ? 1 : 0,
      playsinline: playsInline() ? 1 : 0
    };
  }
  q(command, arg) {
    this.hd({
      event: "command",
      func: command,
      args: arg ? [arg] : void 0
    });
  }
  mc() {
    window.setTimeout(() => this.hd({ event: "listening" }), 100);
  }
  nd(trigger) {
    this.c("loaded-metadata");
    this.c("loaded-data");
    this.b.delegate.kc(void 0, trigger);
  }
  Ba(trigger) {
    this.H?.resolve();
    this.H = null;
    this.c("pause", void 0, trigger);
  }
  Fb(time, trigger) {
    const { duration, realCurrentTime } = this.b.$state, hasEnded = this.nc === YouTubePlayerState.Ve, boundTime = hasEnded ? duration() : time, detail = {
      currentTime: boundTime,
      played: this.Ek(boundTime)
    };
    this.c("time-update", detail, trigger);
    if (!hasEnded && Math.abs(boundTime - realCurrentTime()) > 1) {
      this.c("seeking", boundTime, trigger);
    }
  }
  Ek(time) {
    return this.Ga >= time ? this.Ha : this.Ha = new TimeRange(0, this.Ga = time);
  }
  jc(buffered, seekable, trigger) {
    const detail = {
      buffered: new TimeRange(0, buffered),
      seekable
    };
    this.c("progress", detail, trigger);
    const { seeking, realCurrentTime } = this.b.$state;
    if (seeking() && buffered > realCurrentTime()) {
      this.cb(trigger);
    }
  }
  cb(trigger) {
    const { paused, realCurrentTime } = this.b.$state;
    window.clearTimeout(this.oc);
    this.oc = window.setTimeout(
      () => {
        this.c("seeked", realCurrentTime(), trigger);
        this.oc = -1;
      },
      paused() ? 100 : 0
    );
    this.Wk = false;
  }
  Eb(trigger) {
    const { seeking } = this.b.$state;
    if (seeking())
      this.cb(trigger);
    this.c("pause", void 0, trigger);
    this.c("end", void 0, trigger);
  }
  Sg(state, trigger) {
    const { started, paused, seeking } = this.b.$state, isPlaying = state === YouTubePlayerState.We, isBuffering = state === YouTubePlayerState.Og, isPlay = (paused() || this.F) && (isBuffering || isPlaying);
    if (isBuffering)
      this.c("waiting", void 0, trigger);
    if (seeking() && isPlaying) {
      this.cb(trigger);
    }
    if (!started() && isPlay && this.Wk) {
      this.F?.reject("invalid internal play operation");
      this.F = null;
      if (isPlaying) {
        this.pause();
        this.Wk = false;
      }
      return;
    }
    if (isPlay) {
      this.F?.resolve();
      this.F = null;
      this.c("play", void 0, trigger);
    }
    switch (state) {
      case YouTubePlayerState.Pg:
        this.nd(trigger);
        break;
      case YouTubePlayerState.We:
        this.c("playing", void 0, trigger);
        break;
      case YouTubePlayerState.Ng:
        this.Ba(trigger);
        break;
      case YouTubePlayerState.Ve:
        this.Eb(trigger);
        break;
    }
    this.nc = state;
  }
  jd({ info }, event) {
    if (!info)
      return;
    const { title, intrinsicDuration, playbackRate } = this.b.$state;
    if (isObject(info.videoData) && info.videoData.title !== title()) {
      this.c("title-change", info.videoData.title, event);
    }
    if (isNumber(info.duration) && info.duration !== intrinsicDuration()) {
      if (isNumber(info.videoLoadedFraction)) {
        const buffered = info.progressState?.loaded ?? info.videoLoadedFraction * info.duration, seekable = new TimeRange(0, info.duration);
        this.jc(buffered, seekable, event);
      }
      this.c("duration-change", info.duration, event);
    }
    if (isNumber(info.playbackRate) && info.playbackRate !== playbackRate()) {
      this.c("rate-change", info.playbackRate, event);
    }
    if (info.progressState) {
      const {
        current,
        seekableStart,
        seekableEnd,
        loaded,
        duration: _duration
      } = info.progressState;
      this.Fb(current, event);
      this.jc(loaded, new TimeRange(seekableStart, seekableEnd), event);
      if (_duration !== intrinsicDuration()) {
        this.c("duration-change", _duration, event);
      }
    }
    if (isNumber(info.volume) && isBoolean(info.muted)) {
      const detail = {
        muted: info.muted,
        volume: info.volume / 100
      };
      this.c("volume-change", detail, event);
    }
    if (isNumber(info.playerState) && info.playerState !== this.nc) {
      this.Sg(info.playerState, event);
    }
  }
  I() {
    this.nc = -1;
    this.oc = -1;
    this.Ga = 0;
    this.Ha = new TimeRange(0, 0);
    this.F = null;
    this.H = null;
    this.Wk = false;
  }
}

export { YouTubeProvider };
