// @vidstack/react/dist/dev/player/vidstack-default-components.js
"use client"

export { j as DefaultAirPlayButton, b as DefaultAudioLayout, g as DefaultBufferingIndicator, n as DefaultCaptionButton, v as DefaultChapterTitle, B as DefaultChaptersMenu, y as DefaultControlsSpacer, p as DefaultFullscreenButton, k as DefaultGoogleCastButton, w as DefaultLiveButton, m as DefaultMuteButton, o as DefaultPIPButton, l as DefaultPlayButton, q as DefaultSeekButton, C as DefaultSettingsMenu, x as DefaultTimeGroup, z as DefaultTimeInfo, A as DefaultTimeInvert, s as DefaultTimeSlider, t as DefaultTitle, h as DefaultTooltip, f as DefaultVideoGestures, d as DefaultVideoLargeLayout, c as DefaultVideoLayout, e as DefaultVideoSmallLayout, r as DefaultVolumeSlider } from '../chunks/vidstack-Tvwu66sE.js';
import 'react';
import '../chunks/vidstack-v3pWuBpF.js';
import '../chunks/vidstack-CbEFfR4V.js';
import '../chunks/vidstack-DIXbEOjz.js';
import '../chunks/vidstack-YJpQDvAd.js';
import 'react-dom';
import '../chunks/vidstack-D2VpSFGD.js';
import '../chunks/vidstack-YhK_ZmK3.js';
