// @vidstack/react/dist/prod/player/vidstack-default-components.js
"use client"

export { j as DefaultAirPlayButton, b as DefaultAudioLayout, g as DefaultBufferingIndicator, n as DefaultCaptionButton, v as DefaultChapterTitle, B as DefaultChaptersMenu, y as DefaultControlsSpacer, p as DefaultFullscreenButton, k as DefaultGoogleCastButton, w as DefaultLiveButton, m as DefaultMuteButton, o as DefaultPIPButton, l as DefaultPlayButton, q as DefaultSeekButton, C as DefaultSettingsMenu, x as DefaultTimeGroup, z as DefaultTimeInfo, A as DefaultTimeInvert, s as DefaultTimeSlider, t as DefaultTitle, h as DefaultTooltip, f as DefaultVideoGestures, d as DefaultVideoLargeLayout, c as DefaultVideoLayout, e as DefaultVideoSmallLayout, r as DefaultVolumeSlider } from '../chunks/vidstack-BFRZ2qMl.js';
import 'react';
import '../chunks/vidstack-BNPvxaQh.js';
import '../chunks/vidstack-C023lPlt.js';
import '../chunks/vidstack-BRXztYRL.js';
import '../chunks/vidstack-D2C8-WOh.js';
import 'react-dom';
import '../chunks/vidstack-B-S8hZH2.js';
import '../chunks/vidstack-uileeGqY.js';
